Addressed the preserved-objective replay no-delta idempotency blocker reported on `3753249`; new head is `3e97879`.

What changed:
- Added a regression to replay the assembled preserved-objective active context by itself after restart and assert the durable store does not grow.
- Kept the existing appended-delta regression for `active_context + new user`.
- Kept the inverse guard for repeated user deltas: preserved-objective scaffold replay still does **not** apply when the matching suffix contains a user turn.
- Minimal code change: scaffold suffix replay no longer requires an appended delta (`cursor < len(messages)`), because a pure replay should also be idempotent.

Validation:
- RED before fix: focused preserved-objective replay test failed with `AssertionError: assert 7 == 6`.
- Focused regressions: `2 passed, 28 warnings`.
- `TestAssemblyBudgetSelection`: `4 passed, 28 warnings`.
- Full suite with standalone `agent.context_engine` stub: `817 passed, 248 warnings`.
- Hygiene: `compileall`, `py_compile scripts/import_lossless_claw.py`, `bash -n scripts/install.sh scripts/update.sh`, `git diff --check`, and added-line secret/private-marker scan all clean.

--- a/engine.py
+++ b/engine.py
@@ -3782,7 +3782,7 @@ class LCMEngine(ContextEngine):
                         skipped_tail_gap = True
                         continue
                     break
-                if skipped_tail_gap and kept_tail_reversed:
+                if skipped_tail_gap:
                     break
                 kept_tail_reversed.append(msg)
                 tail_token_total += msg_tokens

--- a/tests/test_lcm_core.py
+++ b/tests/test_lcm_core.py
@@ -3306,6 +3306,37 @@ class TestAssemblyBudgetSelection:
             for msg in assembled
         )
 
+    def test_non_contiguous_raw_user_tail_replay_does_not_duplicate_durable_rows(self, tmp_path, monkeypatch):
+        engine = self._engine(tmp_path, monkeypatch, max_assembly_tokens=160)
+        messages = [
+            {"role": "system", "content": "sys"},
+            {"role": "user", "content": "repeat user intent"},
+            {"role": "assistant", "content": "huge assistant output " * 400},
+        ]
+        engine._ingest_messages(messages)
+        assert engine._store.get_session_count("assembly-session") == len(messages)
+
+        assembled = engine._assemble_context(messages[0], messages[1:])
+        contents = "\n".join(str(msg.get("content", "")) for msg in assembled)
+        assert "repeat user intent" in contents
+        assert "huge assistant output" not in contents
+        assert not any(
+            msg.get("role") == "user" and msg.get("content") == "repeat user intent"
+            for msg in assembled
+        )
+
+        from hermes_lcm.engine import LCMEngine
+
+        replay = LCMEngine(config=engine._config, hermes_home=str(tmp_path / "hermes"))
+        replay._session_id = "assembly-session"
+        replay._ingest_cursor_needs_reconcile = True
+        replay._ingest_messages(assembled + [{"role": "user", "content": "new user after restart"}])
+
+        rows = replay._store.get_session_messages("assembly-session")
+        assert len(rows) == len(messages) + 1
+        assert [row["content"] for row in rows].count("repeat user intent") == 1
+        assert rows[-1]["content"] == "new user after restart"
+
     def test_non_contiguous_preserved_prompt_replay_does_not_duplicate_durable_rows(self, tmp_path, monkeypatch):
         engine = self._engine(tmp_path, monkeypatch, max_assembly_tokens=450)
         messages = [

--- a/tests/test_lcm_engine.py
+++ b/tests/test_lcm_engine.py
@@ -8166,7 +8166,7 @@ class TestAssemblyGuardrails:
         config = LCMConfig(
             fresh_tail_count=10,
             database_path=str(tmp_path / "lcm_guardrail_newest.db"),
-            max_assembly_tokens=50,
+            max_assembly_tokens=120,
         )
         instance = LCMEngine(config=config)
         instance._session_id = "guardrail-session"
@@ -8183,11 +8183,13 @@ class TestAssemblyGuardrails:
             {"role": "system", "content": "s" * 10},
             [
                 {"role": "user", "content": "a" * 20},
-                {"role": "assistant", "content": "b" * 60},
+                {"role": "assistant", "content": "b" * 140},
             ],
         )
 
-        assert [msg["content"] for msg in result[1:]] == ["a" * 20]
+        contents = [msg["content"] for msg in result[1:]]
+        assert any("a" * 20 in content for content in contents)
+        assert not any(msg.get("role") == "user" and msg.get("content") == "a" * 20 for msg in result[1:])
 
     def test_context_anchor_is_budgeted_under_max_assembly_tokens(self, tmp_path, monkeypatch):
         import importlib

